modules/terraform/var-file.go

```diff
@@ -117,33 +102,135 @@ func GetVariableAsListFromVarFileE(t *testing.T, fileName string, key string) ([
+// GetAllVariablesFromVarFileE Parses all data from a provided input file found in VarFile and returns a map of the
+// variable values where the keys are the var names and the value are the parsed var values. Returns an error if the
+// specified file does not exist, the specified file is not readable, or the specified file cannot be decoded from HCL
+func GetAllVariablesFromVarFileE(t *testing.T, fileName string) (map[string]interface{}, error) {
 	fileContents, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return nil, err
+	}
+
+	out, err := parseAndDecodeVarFile(string(fileContents), fileName)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
 
+// parseAndDecodeHcl uses the HCL2 parser to parse the given varfile string into an HCL file body, and then decode it
+// into a map that maps var names to values.
+func parseAndDecodeVarFile(hclContents string, filename string) (out map[string]interface{}, err error) {
+	// The HCL2 parser and especially cty conversions will panic in many types of errors, so we have to recover from
+	// those panics here and convert them to normal errors
+	defer func() {
+		if recovered := recover(); recovered != nil {
+			err = PanicWhileParsingVarFile{RecoveredValue: recovered, ConfigFile: filename}
+		}
+	}()
+
+	parser := hclparse.NewParser()
+
+	file, parseDiagnostics := parser.ParseHCL([]byte(hclContents), filename)
+	if parseDiagnostics != nil && parseDiagnostics.HasErrors() {
+		return nil, parseDiagnostics
+	}
+
+	// VarFiles should only have attributes, so extract the attributes and decode the expressions into the return map.
+	attrs, hclDiags := file.Body.JustAttributes()
+	if hclDiags != nil && hclDiags.HasErrors() {
+		return nil, hclDiags
+	}
+
+	valMap := map[string]cty.Value{}
+	for name, attr := range attrs {
+		val, hclDiags := attr.Expr.Value(nil) // nil because no function calls or variable references are allowed here
+		if hclDiags != nil && hclDiags.HasErrors() {
+			return nil, hclDiags
+		}
+		valMap[name] = val
+	}
+	varMap, err := parseCtyValueMapToGoMap(valMap)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	return varMap, nil
+}
+
+// The following functions were lifted from terragrunt -
+// https://github.com/gruntwork-io/terragrunt/blob/874d2db4618afc7dc8e153ab0219282147865d4a/config/cty_helpers.go
 
-	err = hcl.Decode(out, string(fileContents))
+// This is a hacky workaround to convert a cty Value map to a Go map[string]interface{}. cty does not support this directly
+// (https://github.com/hashicorp/hcl2/issues/108) and doing it with gocty.FromCtyValue is nearly impossible, as cty
+// requires you to specify all the output types and will error out when it hits interface{}. So, as an ugly workaround,
+// we convert the given value to JSON using cty's JSON library and then convert the JSON back to a
+// map[string]interface{} using the Go json library.
+func parseCtyValueMapToGoMap(valMap map[string]cty.Value) (map[string]interface{}, error) {
```

Comment:
Ugh, I'm not a fan of perpetuating this hack. So the `hcl2` lib doesn't have the ability to parse data into arbitrary structs passed in by the user? 

Also, did this function in Terragrunt have unit tests? If so, we prob need to port those over too.

Reply:
The `hcl2` library supports parsing the data to arbitrary structs _except map[string]interface{}_. See the new implementation in f8477eb, where I special case the arbitrary map.

> Also, did this function in Terragrunt have unit tests? If so, we prob need to port those over too.

It looks like this function was implicitly tested in terragrunt. I think the implicit tests we have in here should be sufficient.

Reply:
Roger!